File: give_a_dollar.py
#!/usr/bin/python3
#
# give_a_dollar.py
#
# A Python tool to studying an interesting scenario by Jim McClanahan W4JBM
#
# The Scenario:
#
# 45 people are each given $45. For 5,000 rounds, each person who has money
# must give $1 to another random person (who is, obviously, not themselves).
#
# The question is what does the distribution of money look like over a long
# period of time. Most people (even those familar with statistics and probability)
# tend to guess that the amount each person has will remain relatively flat.
#
# Surprisingly, this is not what happens...
#
# Results are in the format of a GIF. To resize the GIF, you can use the command:
#
# $ gifsicle --resize 320x480 --colors 16 gad.gif > resized.gif
#
# Originally in R from dggoldst at:
# https://gist.github.com/dggoldst/e77cacae87a9b88de52f7868936899c5#file-give_a_dollar-r
#
# Translating R to Python was more difficult than it was worth (since I know
# virtually nothing about R), but some of the variable names survived. :-)
#
# Details on building a gif included info from Thiago Carvalho at:
# https://towardsdatascience.com/basics-of-gifs-with-pythons-matplotlib-54dd544b6f30
#
# Info on merging two images from radarhere at:
# https://stackoverflow.com/questions/53876007/how-to-vertically-merge-two-images
#
# Note: Running under a fairly fresh copy of MX Linux, I would get errors on
# exit after larger runs along the lines of:
#
# ICE default IO error handler doing an exit(), pid=xxx, errno = 32
#
# When I removed the version of matplotlib installed using pip and instead used
# one installated with apt install, the problem went away.

import os
import numpy as np
import matplotlib.pyplot as plt
import random
import imageio.v2 as imageio
from PIL import Image
from tqdm.auto import tqdm

NUMPLAYERS = 45
INITWEALTH = 45
ROUNDS = 5000
GRAPH_MAX = 200

BANK_MIN = INITWEALTH
BANK_MAX = INITWEALTH

filenames = []
filemerge = []

#initialize the bank
#columns wealths of the NUMPLAYERS players
#rows show wealths of each of the ROUNDS ticks of the clocks
bank = np.empty(NUMPLAYERS)
bank.fill(INITWEALTH)

# print(bank)

# This function takes a "player number" as its input and then
# returns a random player number that is not equal to itself.
#
# This is used so that i can give $1 to j.
def give_to(i):
    j = random.choice(list(set([x for x in range(0, NUMPLAYERS)]) - set([i])))
    return(j)

print("Running itterations and creating frames...")

for current_round in tqdm(range(0, ROUNDS+1)):
    for player in range(0, NUMPLAYERS):

        # For Round #1, let's plot the initial state
        if current_round == 0:
            continue ;

        # Does this player have any money left?
        if bank[player] == 0:
            continue ;

        # If so, give someone a dollars
        bank[player] -= 1
        bank[give_to(player)] += 1

    # Capture Min and Max for round
    BANK_MIN = min(np.amin(bank),BANK_MIN)
    BANK_MAX = max(np.amax(bank),BANK_MAX)

    # The following is used for times when I don't need a graph
    # each individual round--% 10 will cut the size of the file
    # and the run time significantly by only graphing 1 in 10
    # rounds

    if (current_round % 1) == 0:

        # Plot the results by Player
        plt.title("Unsorted by Player")
        plt.bar(range(0,NUMPLAYERS), bank, color='b')
        plt.ylim(0,GRAPH_MAX)

        filename = 'plrs' + str(current_round) + '.png'
        filenames.append(filename)

        plt.savefig(filename)
        plt.close()

        # Plot the results Sorted (low to high)
        plt.title("Sorted Distribution")
        plt.bar(range(0,NUMPLAYERS), np.sort(bank), color='r')
        plt.ylim(0,GRAPH_MAX)
        plt.figtext (0.01,0.01,"https://www.github.com/w4jbm/give-a-dollar",horizontalalignment='left')
        plt.figtext (0.99,0.01,current_round,fontfamily="monospace",horizontalalignment='right')


        filename = 'sort' + str(current_round) + '.png'
        filenames.append(filename)

        plt.savefig(filename)
        plt.close()

        # Now merge the two images
        filename = 'merg' + str(current_round) + '.png'
        imgs = [Image.open(i) for i in filenames]

        min_img_width = min(i.width for i in imgs)

        total_height = 0
        for i, img in enumerate(imgs):
            # If the image is larger than the minimum width, resize it
            if img.width > min_img_width:
                imgs[i] = img.resize((min_img_width, int(img.height / img.width * min_img_width)), Image.ANTIALIAS)
            total_height += imgs[i].height

        img_merge = Image.new(imgs[0].mode, (min_img_width, total_height))
        y = 0
        for img in imgs:
            img_merge.paste(img, (0, y))
            y += img.height
        img_merge.save(filename)
        img_merge.close()

        # Now get rid of the two files we had prior to merging things
        # except save the final results
        if current_round != ROUNDS:
            for f in filenames:
                os.remove(f)

        # Save to list of images to go in GIF
        filemerge.append(filename)

        # Clear the working list
        filenames = []

print("Merging frames to create GIF...")

# We have all the files created, so now me merge them!
with imageio.get_writer('gad.gif', mode='I') as writer:
    for f in tqdm(filemerge):
        image = imageio.imread(f)
        writer.append_data(image)
    writer.close()

# Remove files
for f in filemerge:
    os.remove(f)

print()
print("RESULTS SUMMARY:")
print("Over all runs, values ranged from Minimum of",BANK_MIN,"to Maximum of", BANK_MAX)
print()
print("Final values ranged from a Minimum of",np.amin(bank),"to a Maximum of",np.amax(bank))
print("Final Mean was",np.mean(bank),"with a Standard Deviation of",np.std(bank))
print("Final values have a Median of",np.median(bank))

exit()

File: give_a_dollar.jl
#!/usr/bin/julia
#
# give_a_dollar.jl
#
# A Julia tool to studying an interesting scenario by Jim McClanahan W4JBM
#
# The Scenario:
#
# 45 people are each given $45. For 5,000 rounds, each person who has money
# must give $1 to another random person (who is, obviously, not themselves).
#
# The question is what does the distribution of money look like over a long
# period of time. Most people (even those familar with statistics and probability)
# tend to guess that the amount each person has will remain relatively flat.
#
# Surprisingly, this is not what happens...
#
# Individual graphs of the progress are stored and later combined into a GIF.
# I used the Linux convert command to do this. It can be done in Julia but my
# first couple of attempts didn't work, so I fell back to this approach.
#
# I converted my Python version, but one resource I used for that is a version
# in R from dggoldst at:
# https://gist.github.com/dggoldst/e77cacae87a9b88de52f7868936899c5#file-give_a_dollar-r
#
# Translating R to Python was more difficult than it was worth (since I know
# virtually nothing about R), but some of the variable names survived. :-)

using Plots
using Glob

# Initialize Variable
num_players = 45
starting_dollars = 45
num_rounds = 5000

# Figure out number of digits for zero padding
# (I'm not sure the padding is absolutely needed, but it ensures
# the PNG files are in the right order both with the time stamp
# and when sorted by name.)
num_digits = Int(floor(log10(Float64(num_rounds)))) + 1

# Give everyone their $45
bank = fill(starting_dollars, num_players)

# Variable used to capture minimum and maximum values over the series of rounds
global player_min = starting_dollars
global player_max = starting_dollars

# A function that picks a player to give $1 to--but we can't give it
# to ourselves...
function give_to(i, num_players)
    j = rand(1:num_players-1)
    if j == i
        j += 1
    end
    return j
end

for current_round in 0:num_rounds
    for player in 1:num_players

        # If round zero, don't do anything
        # (Let's us put 'base case' in the final GIF
        if current_round == 0
            break
        end

        # Does this player have any money left?
        if bank[player] == 0
            continue
        end
        
        # If we have $1, give it to someone else
        bank[player] -= 1
        bank[give_to(player, num_players)] += 1
        
    end

    # Capture minimum over all rounds
    if (minimum(bank) < player_min)
        global player_min = minimum(bank)
    end

    # Capture maximum over all rounds    
    if (maximum(bank) > player_max)
        global player_max = maximum(bank)
    end

    # This is kind of a kludge, but if we are building a GIF
    # to share, we don't want it to be a huge file. This will
    # let you capture each round (with % 1), every tenth round
    # (with % 10) or other combinations you want.
    #
    # Every tenth for 1,000 rounds results in a GIF that is
    # 4.7 MB in size.
    #
    # NOTE: convert seems to be memory constrained and can only
    # merge a few hundred files without throwing a "resources
    # exhausted" error. Use ffmpeg if you need more frames.
    if (current_round % 25) ≠ 0
        continue
    end

    # Some of the code commented out below lets you generate
    # seperate plots for raw values and the distribution curve
    # instead of just a combined view.

    # Plot the raw values from the bank
    xs = range(1, length=num_players)
    ys = bank
#   pname = string("plot_raw", string(current_round), ".png")
    p1 = bar(xs, ys, ylims=(0,200), xlims=(0.5,45.5))
    ylabel!("Bank Balance")
    xlabel!("Player Number")
    title!("Give-a-Dollar Raw Player Data")
#   savefig(p1, pname)

    # Plot the sorted values (distribution) from the bank
#   xs = range(1, length=num_players)
    ys = sort(bank)
#   pname = string("plot_sorted", string(current_round), ".png")
    p2 = bar(xs, ys, ylims=(0,200), xlims=(0.5,45.5))
    ylabel!("Bank Balance")
    xlabel!(string("Distribution for Round ", current_round))
    title!("Give-a-Dollar Sorted Distribution")
#   savefig(p2, pname)

    pname = string("plot_merged_", lpad(current_round, num_digits, "0"), ".png")
    plot!(size=(800,800))
    final = plot(p1, p2, layout = (2,1), legend=false)
    savefig(final, pname)
    
    
end

# Use Linux commands to merge all the PNGs into a GIF
#
# Two approaches can be used, each with different tradeoffs...

# Using 'convert' gives high quality but will run out of
# buffer space after a few hundred frames and produces large
# file (because quality doesn't come without a cost).
run(`convert -delay 10 -loop 1 \*.png give_a_dollarc.gif`)

# Using 'ffmpeg' reduces the quality because of compression
# but will not overrun a buffer (because results are going to
# the disk) and produces files about half the size of 'convert'.
#run(`ffmpeg -y -framerate 10 -pattern_type glob -i '*.png' -loop -1 give_a_dollarff.gif`)

# Delete all the PNGs now that we're done with them
rm.(glob("plot*.png"))

# Print out some of the details about the run
println("Number of Rounds: ", num_rounds)
println("Minimum Dollars for a Player: ", player_min)
println("Maximum Dollars for a Player: ", player_max)
